### PR TITLE
[1/3] Redfishwrapper changes

### PR DESCRIPTION
**Pull request by joelrebel**
## What does this PR implement/change/remove?

Note: to be merged after https://github.com/bmc-toolbox/bmclib/pull/362

- Adds various helper methods in the `redfishwrapper` for querying Redfish Task, Odata IDs
- Adds generic Redfish methods to Upload firmware and query its status
- Better test coverage

The firmware methods from the [redfish](https://github.com/bmc-toolbox/bmclib/tree/main/providers/redfish) provider will be moved into vendor specific providers,
which will call into the `redfishwrapper` where required. 

Since each vendor requires its own set of OEM parameters and theres differences in the way the Tasks are handled, the current [redfish firmware install methods](https://github.com/bmc-toolbox/bmclib/blob/main/providers/redfish/firmware.go) are getting cumbersome to maintain and extend. 

### Checklist
- [X] Tests added
- [X] Similar commits squashed

### CODE REVIEW

File: internal/redfishwrapper/firmware.go [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+type RedfishUpdateServiceParameters struct {
+	Targets            []string                     `json:"Targets"`
+	OperationApplyTime constants.OperationApplyTime `json:"@Redfish.OperationApplyTime"`
+	Oem                json.RawMessage              `json:"Oem"`
+}
+
+// FirmwareUpload uploads and initiates the firmware install process
+func (c *Client) FirmwareUpload(ctx context.Context, reader io.Reader, params *RedfishUpdateServiceParameters) (taskID string, err error) {
+	// limit to *os.File until theres a need for other types of readers
+	updateFile, ok := reader.(*os.File)
```
Comment (DoctorVin) [inline internal/redfishwrapper/firmware.go:46]: This feels like a mistake.

File: internal/redfishwrapper/client_test.go [hunk 2]
```diff
@@ -83,3 +87,135 @@ func TestWithEtagMatchDisabled(t *testing.T) {
...
+			defer server.Close()
+
+			parsedURL, err := url.Parse(server.URL)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ctx := context.Background()
+
+			//os.Setenv("DEBUG_BMCLIB", "true")
```
Comment (coffeefreak101) [inline internal/redfishwrapper/client_test.go:139]: nit: Commented out code left here

File: internal/redfishwrapper/client.go [hunk 3]
```diff
@@ -74,6 +81,19 @@ func WithEtagMatchDisabled(d bool) Option {
 	}
 }
 
+// WithLogger sets the logger on the redfish wrapper client
+func WithLogger(l *logr.Logger) Option {
+	return func(c *Client) {
+		if l == nil {
+			c.logger = logr.Discard()
```
[the remark this answers is not in the document]
Comment (joelrebel) [reply internal/redfishwrapper/client.go]: Thanks, fixed!

File: internal/redfishwrapper/client_test.go (hunk 2 above)
Comment (joelrebel) [reply internal/redfishwrapper/client_test.go:139]: Thanks, fixed!

File: internal/redfishwrapper/firmware.go (hunk 1 above)
Comment (joelrebel) [reply internal/redfishwrapper/firmware.go:46]: Based on our discussion, I agree and the changes are in the child branch https://github.com/bmc-toolbox/bmclib/pull/363/commits/3a43b9230c14eb9d6fd634f544bfa58b254d47a9

**Comment by joelrebel** [comment]
rebase on main

**Review by DoctorVin — APPROVED** [review]

**Pull request merged**